scripts/bfasst/flows.py

```diff
@@ -61,14 +75,79 @@ def get_flow_fcn_by_name(flow_name):
     return fcn
 
 
-def run_flow(design, flow_type, build_dir, print_to_stdout=True):
+def run_flow(design, flow_type, flow_args, build_dir, print_to_stdout=True):
     assert type(design) is bfasst.design.Design
 
     flow_fcn = flows.get_flow_fcn_by_name(flow_type)
-    return flow_fcn(design, build_dir, print_to_stdout)
-
+    return flow_fcn(design, flow_args, build_dir, print_to_stdout)
+
+
+def parse_flow_args(flow_args):
+    current_state = State.INIT
+    next_state = State.INIT
+
+    args = []
+    args_num = 4
+
+    # Initializing args values
+    for i in range(args_num):
+        args.append("None")
+
+    # Parse flow arguments to get the stages and tools
+    stage = ""
+    tool = ""
+
+    stages = []
+    tools = []
+
+    for i in range(len(flow_args)):
+        current_state = next_state
+
+        # SM for Transitions and Actions
+        if current_state == State.INIT:
+            if flow_args[i] == '-':
+                next_state = State.READING_STAGE
+            elif flow_args[i] == ' ':
+                next_state = State.READING_TOOL
+
+        elif current_state == State.READING_STAGE:
+            if flow_args[i] == '-':
+                next_state = State.READING_STAGE
+            elif flow_args[i] == ' ':
+                stages.append(stage)
+                stage = ""
+                next_state = State.READING_TOOL
+            else:
+                stage = stage + flow_args[i]
+
+        elif current_state == State.READING_TOOL:
+            if flow_args[i] == '-':
+                next_state = State.READING_STAGE
+            elif flow_args[i] == ' ':
+                tools.append(tool)
+                tool = ""
+                next_state = State.READING_TOOL
+            else:
+                tool = tool + flow_args[i]
+                if i == (len(flow_args) - 1):
+                    tools.append(tool)
+                    tool = ""
+                    next_state = State.INIT
+
+    # Getting args from tools based on the stages
+    for i, stage in enumerate(stages):
+        if stage == "synthesis":
+            args[Args.SYNTH_STAGE.value] = tools[i]
+        elif stage == "implementation":
+            args[Args.IMPL_STAGE.value] = tools[i]
+        elif stage == "mapping":
+            args[Args.MAP_STAGE.value] = tools[i]
+        elif stage == "equivalence":
+            args[Args.EQUIV_STAGE.value] = tools[i]
+    
+    return args
 
-def flow_ic2_lse_conformal(design, build_dir):
+def flow_ic2_lse_conformal(design, flow_args, build_dir):
```

Comment:
This is correct, to pass `flow_args` into each flow.  But then within the flow, we want to pass along the option strings into each tool.  

For example:
```
synth_tool = bfasst.synth.ic2_lse.IC2_LSE_SynthesisTool(build_dir, flow_args[SYNTH_STAGE])
```

Then within `tool.py`, at the constructor: https://github.com/byuccl/bfasst/blob/4081d15f33a9a95c4c33e18b84eb8a466a3b0607/scripts/bfasst/tool.py#L19 you would want to accept the argument string for this tool, and use argparse to parse it into a `self.args` that you can then reference for the individual tool.

Does that make sense?

Reply:
Although you can add the `args` argument to the base `Tool` class, the actual logic for the argument parsing will be depending on the type of tool, so the argparse code would actually be placed in the base class for each tool type, such as in the `__init__` of `SynthesisTool`.